test/src/test/java/org/corfudb/runtime/object/transactions/UndoTest.java

```diff
@@ -198,4 +198,62 @@ public void canUndoAfterNoUndo()
         });
 
     }
+
+    /**
+     * Check that rollbackUnsafe can rollback on context switch
+     *
+     * <p>
+     * Verify that if we have two threads with different snapshot we can
+     * Undo the log upon context switch.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void CanUndoDifferentVersion()
+        throws Exception {
+        Map<Integer, String> testMap = getRuntime()
+                .getObjectsView()
+                .build()
+                .setStreamName("test")
+                .setTypeToken(new TypeToken<SMRMap<Integer, String>>() {
+                })
+                .open();
+
+        final String normalValue = "z";
+        final int mapSize = 10 * PARAMETERS.NUM_ITERATIONS_LARGE;
+
+        // populate the map with many elements
+        for (int i = 0; i < mapSize; i++)
+            testMap.put(i, normalValue);
+
+        // t0 is starting a transaction and reading the map
+        // t0 snapshot is at version x (object is at version x)
+        t(0, () -> getRuntime().getObjectsView().TXBegin());
+        t(0, () -> testMap.size());
```

Comment:
This will force a read to the map. Transactional snapshot is lazily initialized upon the first access and size() is a accessor. If we don't read it, we won't have different snapshot for each transaction.